python/lsst/scarlet/lite/image.py

```diff
@@ -0,0 +1,1372 @@
+        self[:] = self.__lshift__(other)
+        return self
+
+    def __rlshift__(self, other: ScalarLike) -> Image:
+        """Shift other to the left by this image bits."""
+        return self.copy_with(data=other << self.data)
+
+    def __rshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the right by other bits."""
+        if not issubclass(np.dtype(type(other)).type, np.integer):
+            raise TypeError("Bit shifting an image can only be done with integers")
+        return self.copy_with(data=self.data >> other)
+
+    def __irshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the right by other bits in place."""
+        self[:] = self.__rshift__(other)
+        return self
+
+    def __rrshift__(self, other: ScalarLike) -> Image:
+        """Shift other to the right by this image bits."""
+        return self.copy_with(data=other >> self.data)
+
+    def __matmul__(self, other: Image | ScalarLike) -> Image:
+        """Matrix multiplication is not supported using images."""
+        raise TypeError("Images do not support matrix mutliplication")
+
+    def __str__(self):
+        """Display the image array, bands, and bounding box."""
+        return f"Image:\n {str(self.data)}\n  bands={self.bands}\n  bbox={self.bbox}"
+
+    def _is_spectral_index(self, index: Any) -> bool:
+        """Check to see if an index is a spectral index.
+
+        Parameters
+        ----------
+        index:
+            Either a slice, a tuple, or an element in `Image.bands`.
+
+        Returns
+        -------
+        result:
+            ``True`` if `index` is band or tuple of bands.
+        """
+        bands = self.bands
+        if isinstance(index, slice):
+            if index.start in bands or index.stop in bands or (index.start is None and index.stop is None):
+                return True
+            return False
+        if index in self.bands:
+            return True
+        if isinstance(index, tuple) and index[0] in self.bands:
+            return True
+        return False
+
+    def _get_box_slices(self, bbox: Box) -> tuple[slice, slice]:
+        """Get the slices of the image to insert it into the overlapping
+        region with `bbox`."""
+        overlap = self.bbox & bbox
+        if overlap != bbox:
+            raise IndexError("Bounding box is outside of the image")
+        origin = bbox.origin
+        shape = bbox.shape
+        y_start = origin[0] - self.yx0[0]
+        y_stop = origin[0] + shape[0] - self.yx0[0]
+        x_start = origin[1] - self.yx0[1]
+        x_stop = origin[1] + shape[1] - self.yx0[1]
+        y_index = slice(y_start, y_stop)
+        x_index = slice(x_start, x_stop)
+        return y_index, x_index
+
+    def _get_sliced(self, indices: Any, value: Image | None = None) -> Image:
+        """Select a subset of an image
+
+        Parameters
+        ----------
+        indices:
+            The indices to select a subsection of the image.
```

Comment:
Ah, yes. This is an example of the docs bitrotting before the PR was even merged. The spatial index is now _required_ to be a `Box`!

Reply:
No, it's a bit more complicated because there are so many different allowable combinations. For example, the index could be a single immutable `band_name`, which can be any immutable object. It could also be a `tuple[band_name, ...])`, or a `tuple[band_name, Box]`, or a `tuple[tuple[band_names, ...], Box]` or a `Box`. But AFAIK, there is no "Immutable" type hint to replace `band_name`, so the `band_name` index really can be `Any`. So there is no advantage to adding other subtypes.